src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,778 @@
+                }
+                double[] filteredIfftMagList = AverageFilter(ifftMagList, Math.Min(ifftMagList.Length, _judgementWindowSize));
+
+                // Step 7: Calculate raw score and set result
+                for (int i = 0; i < results.GetLength(0); ++i)
+                {
+                    var score = CalculateSocre(ifftMagList[i], filteredIfftMagList[i]);
+                    score /= 10.0f;
+                    score = Math.Min(score, 1);
+                    score = Math.Max(score, 0);
+
+                    var detres = score > threshold ? 1 : 0;
+
+                    results[i][0] = detres;
+                    results[i][1] = score;
+                    results[i][2] = ifftMagList[i];
+                }
+            }
+
+            private static double[] BackAdd(double[] data)
+            {
+                double[] predictArray = new double[_lookaheadWindowSize + 1];
+                int j = 0;
+                for (int i = data.Length - _lookaheadWindowSize - 2; i < data.Length - 1; ++i)
+                {
+                    predictArray[j++] = data[i];
+                }
+                var predictedValue = PredictNext(predictArray);
+                double[] backAddArray = new double[data.Length + _backAddWindowSize];
+                for (int i = 0; i < data.Length; ++i)
+                {
+                    backAddArray[i] = data[i];
+                }
+                for (int i = 0; i < _backAddWindowSize; ++i)
+                {
+                    backAddArray[data.Length + i] = predictedValue;
+                }
+                return backAddArray;
+            }
+
+            private static double PredictNext(double[] data)
+            {
+                var n = data.Length;
+                double slopeSum = 0.0f;
+                for (int i = 0; i < n - 1; ++i)
+                {
+                    slopeSum += (data[n - 1] - data[i]) / (n - 1 - i);
+                }
+                return (data[1] + slopeSum);
+            }
+
+            private static double[] AverageFilter(double[] data, int n)
+            {
+                double cumsum = 0.0f;
+                int length = data.Length;
+                double[] cumSumList = new double[length];
+                double[] cumSumShift = new double[length];
+
+                for (int i = 0; i < length; ++i)
+                {
+                    cumsum += data[i];
+                    cumSumList[i] = cumsum;
+                    cumSumShift[i] = cumsum;
+                }
+                for (int i = n; i < length; ++i)
+                {
+                    cumSumList[i] = (cumSumList[i] - cumSumShift[i - n]) / n;
+                }
+                for (int i = 1; i < n; ++i)
+                {
+                    cumSumList[i] /= (i + 1);
+                }
+                return cumSumList;
+            }
+
+            private static double CalculateSocre(double mag, double avgMag)
```

Comment:
>Soc [](start = 43, length = 3)

Typo.

Reply:
Fixed.

---
In reply to: [426435676](https://github.com/dotnet/machinelearning/pull/5135#discussion_r426435676) [](ancestors = 426435676)